### ISSUE TITLE
The hash_code of a finding is not recalculated after the finding is updated

### ISSUE THREAD

**Issue by macedogm** [DefectDojo/django-DefectDojo#3870]
**Bug description**
It seems that the finding's `hash_code` is not recalculated after the finding is updated. It's generated only once, when the finding is created, and then will remain the same even if its fields are updated.

**Steps to reproduce**
1. Go to `System Settings` and enable `Deduplicate findings`.
2. Create one finding in, for example, product 1, engagement 1, test 1. It's `hash_code` will be generated.
3. Create a second finding in the same test as finding 1. It's `hash_code` will be generated.
4. Now update the second finding to be exactly as the first one. All fields that are used to generate the hashcode (as according to https://github.com/DefectDojo/django-DefectDojo/blob/master/dojo/settings/settings.dist.py#L754) must be identical to the fist finding.
5. Even with the updated fields, the `hash_code` of the second finding will not be updated and will remain the same as when it was created.

**Expected behavior**
When any field of a finding, used to generate its hash code, is updated, it's expected that the finding's `hash_code` to be recalculated and possibly change.

**Deployment method** *(select with an `X`)*
- [ ] Kubernetes
- [X] Docker
- [ ] setup.bash / legacy-setup.bash

**Environment information**
 - Operating System: macOS Catalina 10.15.7 (19H114)
 - DefectDojo Commit Message: `[2021-02-15 13:48:05 +0100] 02f4869d: Bump cryptography from 3.4.4 to 3.4.5 (#3863) [ (HEAD -> improve-parser-gitlab_dep_scan, upstream/dev, dev)]`

**Console logs**
```
DEBUG [dojo.specific-loggers.deduplication:2086] Hash_code already computed for finding
```

**Additional context**
The code that checks if the finding has a `hash_code` or not and skip it if it has is in lines https://github.com/DefectDojo/django-DefectDojo/blob/dev/dojo/models.py#L2084-L2088 . Changing the logic from:

```
        if(dedupe_option):
            if (self.hash_code is not None):
                deduplicationLogger.debug("Hash_code already computed for finding")
            else:
                self.hash_code = self.compute_hash_code()
```

To:

```
        if(dedupe_option):
            self.hash_code = self.compute_hash_code()
```

Will result in 3 tests from `dojo/unittests/test_deduplication_logic.py` to fail.

```
uwsgi_1         | ======================================================================
uwsgi_1         | FAIL: test_dedupe_not_inside_engagement_same_hash_unique_id_or_hash_code2 (dojo.unittests.test_deduplication_logic.TestDuplicationLogic)
uwsgi_1         | ----------------------------------------------------------------------
uwsgi_1         | Traceback (most recent call last):
uwsgi_1         |   File "/app/dojo/unittests/test_deduplication_logic.py", line 798, in test_dedupe_not_inside_engagement_same_hash_unique_id_or_hash_code2
uwsgi_1         |     self.assert_finding(finding_new, not_pk=22, duplicate=True, duplicate_finding_id=124, hash_code=finding_22.hash_code)
uwsgi_1         |   File "/app/dojo/unittests/test_deduplication_logic.py", line 1355, in assert_finding
uwsgi_1         |     self.assertEqual(finding.duplicate_finding.id, duplicate_finding_id)
uwsgi_1         | AssertionError: 224 != 124
uwsgi_1         | 
uwsgi_1         | ======================================================================
uwsgi_1         | FAIL: test_duplicate_after_modification (dojo.unittests.test_deduplication_logic.TestDuplicationLogic)
uwsgi_1         | ----------------------------------------------------------------------
uwsgi_1         | Traceback (most recent call last):
uwsgi_1         |   File "/app/dojo/unittests/test_deduplication_logic.py", line 1198, in test_duplicate_after_modification
uwsgi_1         |     self.assert_finding(finding_new, not_pk=24, duplicate=False, not_hash_code=None)
uwsgi_1         |   File "/app/dojo/unittests/test_deduplication_logic.py", line 1348, in assert_finding
uwsgi_1         |     self.assertEqual(finding.duplicate, duplicate)
uwsgi_1         | AssertionError: True != False
uwsgi_1         | 
uwsgi_1         | ======================================================================
uwsgi_1         | FAIL: test_hash_code_onetime (dojo.unittests.test_deduplication_logic.TestDuplicationLogic)
uwsgi_1         | ----------------------------------------------------------------------
uwsgi_1         | Traceback (most recent call last):
uwsgi_1         |   File "/app/dojo/unittests/test_deduplication_logic.py", line 1167, in test_hash_code_onetime
uwsgi_1         |     self.assertEqual(finding_new.hash_code, hash_code_at_creation)
uwsgi_1         | AssertionError: '3c6f54b701b3ffaf6e16bf5cb709deba190da00c44ce0d1229741fbb03045caa' != '8cba854f0b70f8a25064952402fbe30c728c0017b83c245d786366956044e0bf'
uwsgi_1         | - 3c6f54b701b3ffaf6e16bf5cb709deba190da00c44ce0d1229741fbb03045caa
uwsgi_1         | + 8cba854f0b70f8a25064952402fbe30c728c0017b83c245d786366956044e0bf
uwsgi_1         | 
uwsgi_1         | 
uwsgi_1         | ----------------------------------------------------------------------
```

Looking at the failed tests it's possible to see comments regarding this as a possible bug:
```
# howwever defect dojo does NOT recalculate the hash_code, so it will not mark this finding as duplicate. feature or BUG?
```

**Comment by macedogm** [comment]
Please let me know your thoughts and if this is an expected behaviour or not.

**Comment by valentijnscholten** [comment]
I captured this behaviour in the unit tests to make sure it either stays as it is, or we make a conscious decision to change it and update the tests. My first instinct would say it should be recalculated, but it could lead to situations that a finding then suddenly becomes a duplicate of another finding. Possibly it becomes a duplicate of a newer finding, what could be confusing.

**Comment by macedogm** [comment]
> but it could lead to situations that a finding then suddenly becomes a duplicate of another finding. Possibly it becomes a duplicate of a newer finding, what could be confusing.

I agree with you. This will seem a bit strange for the user and lead to confusion indeed.

Maybe a better behavior would be to some sort flag the finding as a potential duplicate of another finding and display a different message, in order to distinguish from the normal deduplication process. However, I feel that this will led to a more complex logic and behavior instead of a simple one.

In this case, I'm happy to close the ticket if you don't see any other way of improvement.
